docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+```cs
+void CastingReferences(Span<char> buffer)
+{
+    ref char charSearchSpace = ref MemoryMarshal.GetReference(buffer);
+    ref short searchSpace = ref Unsafe.As<char, short>(ref charSearchSpace);
+    // from now on we can use Vector128<short> or Vector256<short>
+}
+```
+
+Or [MemoryMarshal.Cast<TFrom, TTo>](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.memorymarshal.cast#system-runtime-interopservices-memorymarshal-cast-2(system-readonlyspan((-0)))), which casts a span of one primitive type to a span of another primitive type:
+
+```cs
+void CastingSpans(Span<char> chars)
+{
+    Span<short> shorts = MemoryMarshal.Cast<char, short>(chars);
+}
+```
+
+It's also possible to get managed references from unmanaged pointers:
+
+```cs
+void PointerToReference(char* pUtf16Buffer, byte* pAsciiBuffer)
+{
+    // of the same type:
+    ref byte asciiBuffer = ref *pAsciiBuffer;
+    // of different types:
+    ref ushort utf16Buffer = ref *(ushort*)pUtf16Buffer;
+}
+```
+
+It's only safe to convert a managed reference to a pointer if it's known that the reference is already pinned. If it's not, the moment after you get the pointer it could be invalid.
+
+## Mindset
+
+Vectorizing real-world algorithms seems complex at the beginning. And what do software engineers do with complex problems? We break them down into sub-problems until these become simple enough to be solved directly.
+
+Let's implement a vectorized method for checking whether a given byte buffer consists only from valid ASCII characters to see how similar problems can be solved.
+
+### Edge cases
+
+Before we start working on the implementation, let's list all edge cases for our `IsAcii(ReadOnlySpan<byte> buffer)` method (and ideally write tests):
+
+* It does not need to throw any argument exceptions, as `ReadOnlySpan` is `struct` and it can never be `null` or invalid.
+* It should return `true` for an empty buffer.
+* It should detect invalid characters in the entire buffer, including the remainder.
+* It should not read any bytes that don't belong to the provided buffer.
+
+### Scalar solution
+
+Once we know all edge cases, we need to understand our problem and find a scalar solution.
+
+ASCII characters are values in the range from `0` to `127` (inclusive). It means that we can find invalid ASCII bytes by just searching for values that are larger than `127`. If we treat `byte` (unsigned, range from 0 to 255) as `sbyte` (signed, range from -128 to 127), it's a matter of performing "is less than zero" check.
+
+The binary representation of 0-127 range is following:
+
+```log
+00000000
+01111111
+^
+most significant bit
+```
+
+When we look at it, we can realize that another way is checking whether the most significant bit is equal `1`. For the scalar version, we could perform a logical AND:
+
+```cs
+bool IsValidAscii(byte c) => (c & 0b1000_0000) == 0;
+```
+
+### Vectorized solution
```

Comment:
You never finish the vectorized IsAscii implementation. There should be a complete `static bool IsAscii(ReadOnlySpan<byte>)` or similar method that does everything the "right" way.